src/QirRuntime/lib/QIR/bridge-qis.ll

```diff
@@ -29,122 +29,264 @@
-declare void @quantum__qis__x(%class.QUBIT*)
-declare void @quantum__qis__y(%class.QUBIT*)
-declare void @quantum__qis__z(%class.QUBIT*)
-
-declare void @quantum__qis__crx(%struct.QirArray*, double, %class.QUBIT*)
-declare void @quantum__qis__crz(%struct.QirArray*, double, %class.QUBIT*)
+
+declare void @quantum__qis__exp__body(%struct.QirArray*, double, %struct.QirArray*)
+declare void @quantum__qis__exp__adj(%struct.QirArray*, double, %struct.QirArray*)
+declare void @quantum__qis__exp__ctl(%struct.QirArray*, %struct.QirArray*, double, %struct.QirArray*)
+declare void @quantum__qis__exp__ctladj(%struct.QirArray*, %struct.QirArray*, double, %struct.QirArray*)
+declare void @quantum__qis__h__body(%class.QUBIT*)
+declare void @quantum__qis__h__ctl(%struct.QirArray*, %class.QUBIT*)
+declare %class.RESULT* @quantum__qis__measure__body(%struct.QirArray*, %struct.QirArray*)
+declare void @quantum__qis__r__body(i32, double, %class.QUBIT*)
+declare void @quantum__qis__r__adj(i32, double, %class.QUBIT*)
+declare void @quantum__qis__r__ctl(%struct.QirArray*, i32, double, %class.QUBIT*)
+declare void @quantum__qis__r__ctladj(%struct.QirArray*, i32, double, %class.QUBIT*)
+declare void @quantum__qis__s__body(%class.QUBIT*)
+declare void @quantum__qis__s__adj(%class.QUBIT*)
+declare void @quantum__qis__s__ctl(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__s__ctladj(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__t__body(%class.QUBIT*)
+declare void @quantum__qis__t__adj(%class.QUBIT*)
+declare void @quantum__qis__t__ctl(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__t__ctladj(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__x__body(%class.QUBIT*)
+declare void @quantum__qis__x__ctl(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__y__body(%class.QUBIT*)
+declare void @quantum__qis__y__ctl(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__z__body(%class.QUBIT*)
+declare void @quantum__qis__z__ctl(%struct.QirArray*, %class.QUBIT*)
 
 ;===============================================================================
 ; quantum.qis namespace implementations
 ;
-define double @__quantum__qis__intAsDouble(i64 %i)
-{
-  %d = call double @quantum__qis__intAsDouble(i64 %i)
-  ret double %d
+
+define void @__quantum__qis__exp__body(%Array* %.paulis, double %angle, %Array* %.qubits) {
+  %paulis = bitcast %Array* %.paulis to %struct.QirArray*
+  %qubits = bitcast %Array* %.qubits to %struct.QirArray*
+  call void @quantum__qis__exp__body(%struct.QirArray* %paulis, double %angle, %struct.QirArray* %qubits)
+  ret void
+}
+
+define void @__quantum__qis__exp__adj(%Array* %.paulis, double %angle, %Array* %.qubits) {
+  %paulis = bitcast %Array* %.paulis to %struct.QirArray*
+  %qubits = bitcast %Array* %.qubits to %struct.QirArray*
+  call void @quantum__qis__exp__adj(%struct.QirArray* %paulis, double %angle, %struct.QirArray* %qubits)
+  ret void
 }

```

Comment:
I feel something that I don't know yet behind the observation that all the parameters start with dot `.` (after `%`), but the parameter `double %angle` does not (it has no `.` after `%`).

Reply:
Ah! That's my own sneaky naming convention! :D If you use temporary variables (like `%0` or `%42`) in LLVM's IR, Clang insists on them being in order and having no gaps which makes it a pain to update the code. So I went with naming the locals explicitly, using leading dot (which is _not_ a special char in LLVM's IR) for types from QIR and no dot for native types.

Reply:
I guess I have an intuitive guess of what that means... No need to reply.